sdc/native/hashmap.hpp

```diff
@@ -0,0 +1,907 @@
+    // TO-DO: support copying for all hashmaps and ConcurrentDict's .copy() method in python?
+    NumericHashmapType(const NumericHashmapType&) = delete;
+    NumericHashmapType& operator=(const NumericHashmapType&) = delete;
+    NumericHashmapType(NumericHashmapType&& rhs) : p_map(std::move(rhs->p_map)) {};
+    NumericHashmapType& operator=(NumericHashmapType&& rhs) {
+        this->p_map.reset(std::move(rhs->p_map));
+        return *this;
+    };
+    ~NumericHashmapType() {}
+
+    uint64_t size() {
+        return this->p_map->size();
+    }
+
+    void set(Key key, Val val) {
+        (*this->p_map)[key] = val;
+    }
+
+    int8_t contains(Key key) {
+        auto it = this->p_map->find(key);
+        return it != this->p_map->end();
+    }
+    int8_t lookup(Key key, Val* res) {
+        auto it = this->p_map->find(key);
+        bool found = it != this->p_map->end();
+        if (found)
+            *res = (*it).second;
+
+        return found;
+    }
+    void clear() {
+        this->p_map->clear();
+    }
+
+    int8_t pop(Key key, Val* res) {
+        auto node_handle = this->p_map->unsafe_extract(key);
+        auto found = !node_handle.empty();
+        if (found)
+            *res = node_handle.mapped();
+
+        return found;
+    }
+
+    void update(const NumericHashmapType& other) {
+        this->p_map->merge(*other.p_map);
+    }
+
+    void* getiter() {
+        auto p_it = new iterator_type(this->p_map->begin());
+        auto state = new iter_state((void*)p_it, (void*)this);
+        return state;
+    }
+};
+
+
+template<typename Key,
+         typename Val,
+         typename HashCompare=tbb::tbb_hash_compare<Key>
+>
+class GenericHashmapBase {
+public:
+    using map_type = typename tbb::concurrent_hash_map<Key, Val, HashCompare>;
+    using iterator_type = typename map_type::iterator;
+    std::unique_ptr<map_type> p_map;
+
+    // FIXME: 0 default size is suboptimal, can we optimize this?
+    GenericHashmapBase() : p_map(std::unique_ptr<map_type>(new map_type(0, HashCompare()))) {}
+    GenericHashmapBase(const HashCompare& hash_compare) {
+        this->p_map.reset(new map_type(0, hash_compare));
+    }
+    GenericHashmapBase(std::unique_ptr<map_type>&& rhs_p_map) : p_map(std::move(rhs_p_map)) {};
+
+    GenericHashmapBase(const GenericHashmapBase&) = delete;
+    GenericHashmapBase& operator=(const GenericHashmapBase&) = delete;
+    GenericHashmapBase(GenericHashmapBase&& rhs) : p_map(std::move(rhs->p_map)) {};
+    GenericHashmapBase& operator=(GenericHashmapBase&& rhs) {
+        this->p_map.reset(std::move(rhs->p_map));
+        return *this;
+    };
+    virtual ~GenericHashmapBase() {
+    }
+
+    uint64_t size() {
+        return this->p_map->size();
+    }
+
+    int8_t contains(Key key) {
+        bool found = false;
+        {
+            typename map_type::const_accessor result;
+            found = this->p_map->find(result, key);
+            result.release();
```

Comment:
Do we really need both - inner scope and call to `release`?

Reply:
No of course, but it's (I guess) usual practice to release immediately after you don't need the lock (i.e. I copy-pasted it from TBB examples :)).